R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+  list(package = package, dependencies = dependencies_from_description(package, dcf))
+}
+
+#' Download the accurate remote DESCRIPTION file for a github repo.
+get_remote_dependencies.github <- function(package) {
+  output <- download_description_github(package)
+  list(package = output$package
+    , dependencies = dependencies_from_description(package, output$dcf))
+}
+
+download_description_github <- function(package) {
+  remote <- package$remote
+  filepath <- download_package(structure(
+    package,
+    class = c(remote, class(package))))
+  package$download_path <- filepath
+  file_list <- unzip(filepath, list = TRUE)
+  sep <- .Platform$file.sep
+  subdir <- ""
+  if (!is.null(package$subdir)){
+    subdir <- paste0(sep, package$subdir)
+  }
+  description_name <- file_list$Name[grepl(paste0("^[^", sep, "]+"
+    , subdir, sep, "DESCRIPTION$"), file_list$Name)]
+  file_con <- unz(filepath, description_name)
+  dcf <- read.dcf(file = file_con)
+  close(file_con)
+  list(package = package, dcf = dcf)
+}
+
+version_from_remote <- function(package) {
+  output <- download_description_github(package)
+  unlink(package$download_path)
+  version_from_description(package
+    , output$dcf)
+}
+
+version_from_description <- function(package_name, dcf) {
+  if (!"Version" %in% colnames(dcf)) return(NA)
+  as.character(dcf[1, which(colnames(dcf) == "Version")])
+}
+
+#' Parse dependencies from description
+dependencies_from_description <- function(package, dcf) {
+  ## We install 4 kinds of dependencies listed in the description file. If our
+  ## dcf does not contain any of these elements we have no dependencies to
+  ## speak of
+  dependency_levels <- c("Depends", "Imports", "LinkingTo", "Remotes")
+  if (!any(dependency_levels %in% colnames(dcf))) return(list())
+
+  ## The parse_dcf function returns a matrix where rows correspond to packages
+  ## and column 1 corresponds to the package name and column 3 corresponds to
+  ## its version requirement.  The rownames of this matrix are the type of 
+  ## dependency (Depends, Imports, LinkingTo, or Remotes)
+  dependencies_parsed <- as.data.frame(parse_dcf(dcf
+    , depLevel = dependency_levels[dependency_levels %in%
+      colnames(dcf)])[[package$name]])
+
+  ## We separate out non-remote dependencies from remote dependencies, because
+  ## they require different logic
+  non_remote_dependencies <- dependencies_parsed[rownames(dependencies_parsed) != "Remotes", ]
+  remote_dependencies <- dependencies_parsed[rownames(dependencies_parsed) == "Remotes", ]
+
+  ## Remove the Depends entry that just corresponds to the R version 
+  ## requirements
+  non_remote_dependencies <- non_remote_dependencies[!grepl("^[rR]$"
+    , non_remote_dependencies[,1]), , drop = FALSE]
+
+  ## Parse the non-remote dependencies into a list of packages
+  if (identical(nrow(non_remote_dependencies),0L)){
+    non_remote_list <- list()
+  } else {
+    non_remote_list <- lapply(seq_along(non_remote_dependencies[,1])
```

Comment:
Yes. There you go

On Mon, Feb 29, 2016 at 5:18 PM, davidboren <notifications@github.com>
wrote:

> In R/dependecies.R
> <https://github.com/robertzk/lockbox/pull/73#discussion_r54497329>:
>
> > +
> > +  ## We separate out non-remote dependencies from remote dependencies, because
> > +  ## they require different logic
> > +  non_remote_dependencies <- dependencies_parsed[rownames(dependencies_parsed) != "Remotes", ]
> > +  remote_dependencies <- dependencies_parsed[rownames(dependencies_parsed) == "Remotes", ]
> > +
> > +  ## Remove the Depends entry that just corresponds to the R version
> > +  ## requirements
> > +  non_remote_dependencies <- non_remote_dependencies[!grepl("^[rR]$"
> > +    , non_remote_dependencies[,1]), , drop = FALSE]
> > +
> > +  ## Parse the non-remote dependencies into a list of packages
> > +  if (identical(nrow(non_remote_dependencies),0L)){
> > +    non_remote_list <- list()
> > +  } else {
> > +    non_remote_list <- lapply(seq_along(non_remote_dependencies[,1])
>
> You know what, it's a friggin map:
>
> non_remote_list <- Map(function(u,v) list(name = u, version = v)
>   , as.character(non_remote_dependencies[i,1]), as.character(non_remote_dependencies[i,3]))
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/robertzk/lockbox/pull/73/files#r54497329>.
>



-- 
-Robert Krzyzanowski